Add WTF and bmalloc modulemaps

adetaylor opened the pull request:
#### 305f3c2f21d1f78056ac7a030fd44eb2098d7da2
<pre>
Add WTF and bmalloc modulemaps
<a href="https://bugs.webkit.org/show_bug.cgi?id=292791">https://bugs.webkit.org/show_bug.cgi?id=292791</a>
<a href="https://rdar.apple.com/151023888">rdar://151023888</a>

Reviewed by Yusuke Suzuki and Elliott Williams.

This commit adds modulemap files for WTF and bmalloc. These are necessary
because Swift/C++ interop interprets C++ headers using the &quot;-fmodules&quot; option,
interpreting the C++ codebase as clang modules.

Ultimately this PR results in just two new module.modulemap
files alongside the headers. However, this can cause build time
differences in all the transitive dependencies of WTF and bmalloc
because they&apos;ll interpret header files differently.

The approach taken for the two modulemaps is different.

* bmalloc has a hard-coded modulemap, defining a single monolithic module.
* WTF has an autogenerated modulemap, where each header typically
  resides in a small submodule.

The approach of many small submodules was previously used in
WebKit_Internal, and is used here because Swift/C++ interoperability otherwise
tries to interpret the whole module, and that results in errors such as:

  /usr/include/c++/v1/__memory/construct_at.h:47:58: note: candidate template ignored: substitution failure [with _Tp = WTF::TextBreakIterator, _Args = &lt;const WTF::TextBreakIterator &amp;&gt;]: call to deleted constructor of &apos;WTF::TextBreakIterator&apos;
_LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 _Tp* __construct_at(_Tp* __location, _Args&amp;&amp;... __args) {

This is presumed to be <a href="https://rdar.apple.com/148760413">rdar://148760413</a> or related overinstantiation
of templates. We&apos;ll likely need to solve those in future, but meanwhile
splitting the module into lots of submodules limits the blast radius of
such issues and allows us to make progress with Swift.

Why is one manually generated and the other autogenerated?
Per the comments in the bmalloc modulemap, we only want to expose
a hand-curated set of headers in the bmalloc module, and therefore
there&apos;s no benefit to autogenerating it. It&apos;s specifically the
dependency on a certain value of OS_UNFAIR_LOCK_INLINE which
forces this decision, and that&apos;s discussed in
<a href="https://rdar.apple.com/153287512">rdar://153287512</a>.

Whereas, for WTF we want to expose all the headers, and thus evolving the
modulemap to add new headers and remove older headers would be error-prone and
time consuming - hence we autogenerate.

For the WTF mdoulemap, we use a script and a TOML file that describes the
quirks of given header files (for example, whether they&apos;re interdependent or
can be used in plain C contexts). See the TOML files in this PR and their
comments for a description of these quirks. Thanks to Elliott Williams for
writing the script.

The script converts a tapi filelist-style header listing
into a modulemap, supporting both monolithic and submodule layouts.
This process works well for libraries like WTF which have no
umbrella header, and which have no concept of public vs private API.

We do not export the modulemaps to the build output on pre-OS X 15;
doing so confuses the build of PALSwift.

This commit also adjusts some headers to be more compatible with
modules, adding missing includes, and ensuring all includes are
outside of any extern &quot;C&quot; blocks.

This PR is the second attempt at landing modulemaps for WTF and bmalloc.
The last time didn&apos;t work because it turned out that we needed to resolve
three issues first, to unbreak downstream consumers:

* We needed to move away from reliance on UCHAR_TYPE in WTF.
  (<a href="https://rdar.apple.com/153817245">rdar://153817245</a>).
* Downstream consumers of WebKit were previously using some WTF definitions
  without including WTF headers. That ceases to work when WTF becomes a
  module; we have removed that dependency (<a href="https://rdar.apple.com/154081231">rdar://154081231</a>)
* #45625 resolves an issue with AppServerSupport which starts to appear
  once a modulemap is in place. There were also small fixes to the
  modularization of various system dependencies included within WebKit.
  (Because some of these fixes in WebKit&apos;s dependencies have only landed
  in recent SDKs, we do not export the modulemaps when older SDKs
  are in use.)

* Source/WTF/Configurations/WTF.xcconfig:
* Source/WTF/Configurations/modulemap.toml: Added.
* Source/WTF/Scripts/modulemap-from-tapi-filelist.py: Added.
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/Compiler.h:
* Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:
* Source/WebKit/WebKitSwift/IdentityDocumentServices/WKIdentityDocumentPresentmentDelegate.h:
* Source/bmalloc/Configurations/bmalloc.xcconfig:
* Source/bmalloc/Configurations/module.modulemap: Added.
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:

Co-authored-by: Elliott Williams &lt;emw@apple.com&gt;
Canonical link: <a href="https://commits.webkit.org/297366@main">https://commits.webkit.org/297366@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e1b26cc6c480d9cd75bfe62d1dbbd432b0be261d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111424 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31090 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21555 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117456 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61692 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113386 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31771 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39672 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84683 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114371 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25370 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100307 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65130 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24716 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18443 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61276 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/103917 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94757 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18509 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120673 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/109978 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38473 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28593 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93612 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38849 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96576 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93431 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23824 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38541 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16316 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34507 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38362 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43839 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/134253 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38027 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36184 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41360 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39729 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (changes requested):
I don't think having massive WTF.modulemap etc. is the right solution. We will not update it every time we add / modify WTF, that's significant development regression.

emw-apple commented on Source/WebCore/PAL/Configurations/PAL.xcconfig:
```diff
@@ -71,7 +71,7 @@ SWIFT_OBJC_INTERFACE_HEADER_NAME = PALSwift-Generated.h;
 SWIFT_OBJC_INTEROP_MODE = objcxx;
 
-OTHER_SWIFT_FLAGS = $(OTHER_SWIFT_FLAGS_$(WK_PLATFORM_NAME));
+OTHER_SWIFT_FLAGS = $(OTHER_SWIFT_FLAGS_$(WK_PLATFORM_NAME)) -Xcc -DUCHAR_TYPE=char16_t;
```
Might be preferable to use the native build setting:
```
GCC_PREPROCESSOR_DEFINITIONS = $(inherited) UCHAR_TYPE=char16_t;
```

and also to add it to CommonBase.xcconfig, so we don't have to worry about keeping this in sync as we add Swift-C++ code in more places.

adetaylor replied:
I thought I'd tried that, and had run into problems with it being applied even to C language files where `char16_t` is not a thing, but I just tried it again and haven't hit any problems. So let's do as you suggest for now, in the knowledge that we might need to change it back again in future.

adetaylor replied:
Yes - [here's the doom](https://ews-build.webkit.org/#/builders/131/builds/25376) from C files seeing that definition. I'll add a comment explaining this.

Constellation commented on Source/WTF/Configurations/modulemap.toml:
```diff
@@ -0,0 +1,27 @@
+attributes = ['extern_c']
+requirements = []
+[module.HashMap]
+headers = ['HashMap.h', 'RobinHoodHashTable.h']
```
Can you add some comments there about why and when we need to define these things?

Constellation commented on Source/bmalloc/Configurations/modulemap.toml:
```diff
@@ -0,0 +1,10 @@
+attributes = ['extern_c']
+headers = ['BExport.h']
+
+[module.pas_allocation_mode]
```
Ditto about it

Constellation replied:
Ditto, please add comments about why we need this separately, and when we need these things.

adetaylor replied:
I'll move the comment "can be used in plain C contexts" down a few lines to make it more obvious that it applies to this line.

Constellation commented on Source/WTF/Configurations/modulemap.toml:
```diff
@@ -0,0 +1,43 @@
+requirements = []
+[module.HashMap]
+# These two headers are interdependent so need to be colocated
+# in one submodule.
```
What does it mean? There are many header dependencies, for example, Vector is relying on FastMalloc.h. But they are not having one module for them. So what is the difference?

adetaylor replied:
I'll elaborate on this when I come back to this PR. The key part is _inter_dependent - that is, they depend on each other in a circular fashion.

emw-apple commented on Source/bmalloc/Configurations/module.modulemap:
```diff
@@ -0,0 +1,49 @@
+    // List of "semi-public" bmalloc headers. What does "semi-public" mean? These are
+    // the headers which, in practice, are included by WTF headers. WTF headers are
+    // exported as part of the WebKit API, and therefore these headers need to be
+    // buildable as a clang module.
```
Oh good, I'm glad this worked out!

emw-apple commented on Source/bmalloc/libpas/src/libpas/pas_darwin_spi.h:
```diff
@@ -85,6 +85,14 @@ VM_FLAGS_ALIAS_MASK);
+
+#if defined(__has_include) && __has_include(<stack_logging.h>)
+#include <stack_logging.h>
+#else
```
This is the one snippet I'm not sure I understand—is this some header that the code was getting implicitly from elsewhere?

adetaylor replied:
Comment added!

emw-apple reviewed (commented):
Really appreciate the detailed message. This looks good to me, though I won't r+ while it's in draft in case you have other changes planned.

For the sake of clarity for reviewers, you might want to add me as a coauthor?
```
Co-authored-by: Elliott Williams <emw@apple.com>
```
at the bottom of the commit message should make GitHub show it. WebKit doesn't have official policies on coauthorship, though.

(For other reviewers: I worked on the build script that reads `modulemap.toml` and the Xcode integration.)

emw-apple commented on Source/bmalloc/Configurations/bmalloc.xcconfig:
```diff
@@ -21,6 +21,13 @@
+// affects how downstream components build, so we don't yet enable it
+// by default.
+EXCLUDED_SOURCE_FILE_NAMES_ = module.modulemap;
+EXCLUDED_SOURCE_FILE_NAMES = $(inherited) $(EXCLUDED_SOURCE_FILE_NAMES_$(MODULARIZE_WTF_AND_BMALLOC));
```
Per our discussion, I think we'll want to remove the `MODULARIZE_WTF_AND_BMALLOC` setting, make this behavior unconditional, and wait to land on main until it's safe to do so. Especially since emitting the modulemap does not properly invalidate incremental builds.

emw-apple commented on Source/bmalloc/Configurations/module.modulemap:
```diff
@@ -0,0 +1,55 @@
+    requires cplusplus11
+    // List of "semi-public" bmalloc headers. What does "semi-public" mean? These are
+    // the headers which, in practice, are included by WTF headers. WTF headers are
+    // exported as part of the WebKit API, and therefore these headers need to be
```
nit: To be more precise, I would say "exported as part of WebKit **SPI**". WTF isn't exposed to public API clients.

Constellation reviewed (approved):
One question. What happen if we missed adding headers etc. into these modulemap? Can we get a compile error etc.?

adetaylor committed:
Add WTF and bmalloc modulemaps
https://bugs.webkit.org/show_bug.cgi?id=292791
rdar://151023888

Reviewed by Yusuke Suzuki and Elliott Williams.

This commit adds modulemap files for WTF and bmalloc. These are necessary
because Swift/C++ interop interprets C++ headers using the "-fmodules" option,
interpreting the C++ codebase as clang modules.

Ultimately this PR results in just two new module.modulemap
files alongside the headers. However, this can cause build time
differences in all the transitive dependencies of WTF and bmalloc
because they'll interpret header files differently.

The approach taken for the two modulemaps is different.

* bmalloc has a hard-coded modulemap, defining a single monolithic module.
* WTF has an autogenerated modulemap, where each header typically
  resides in a small submodule.

The approach of many small submodules was previously used in
WebKit_Internal, and is used here because Swift/C++ interoperability otherwise
tries to interpret the whole module, and that results in errors such as:

  /usr/include/c++/v1/__memory/construct_at.h:47:58: note: candidate template ignored: substitution failure [with _Tp = WTF::TextBreakIterator, _Args = <const WTF::TextBreakIterator &>]: call to deleted constructor of 'WTF::TextBreakIterator'
_LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 _Tp* __construct_at(_Tp* __location, _Args&&... __args) {

This is presumed to be rdar://148760413 or related overinstantiation
of templates. We'll likely need to solve those in future, but meanwhile
splitting the module into lots of submodules limits the blast radius of
such issues and allows us to make progress with Swift.

Why is one manually generated and the other autogenerated?
Per the comments in the bmalloc modulemap, we only want to expose
a hand-curated set of headers in the bmalloc module, and therefore
there's no benefit to autogenerating it. It's specifically the
dependency on a certain value of OS_UNFAIR_LOCK_INLINE which
forces this decision, and that's discussed in
rdar://153287512.

Whereas, for WTF we want to expose all the headers, and thus evolving the
modulemap to add new headers and remove older headers would be error-prone and
time consuming - hence we autogenerate.

For the WTF mdoulemap, we use a script and a TOML file that describes the
quirks of given header files (for example, whether they're interdependent or
can be used in plain C contexts). See the TOML files in this PR and their
comments for a description of these quirks. Thanks to Elliott Williams for
writing the script.

The script converts a tapi filelist-style header listing
into a modulemap, supporting both monolithic and submodule layouts.
This process works well for libraries like WTF which have no
umbrella header, and which have no concept of public vs private API.

We do not export the modulemaps to the build output on pre-OS X 15;
doing so confuses the build of PALSwift.

This commit also adjusts some headers to be more compatible with
modules, adding missing includes, and ensuring all includes are
outside of any extern "C" blocks.

This PR is the second attempt at landing modulemaps for WTF and bmalloc.
The last time didn't work because it turned out that we needed to resolve
three issues first, to unbreak downstream consumers:

* We needed to move away from reliance on UCHAR_TYPE in WTF.
  (rdar://153817245).
* Downstream consumers of WebKit were previously using some WTF definitions
  without including WTF headers. That ceases to work when WTF becomes a
  module; we have removed that dependency (rdar://154081231)
* #45625 resolves an issue with AppServerSupport which starts to appear
  once a modulemap is in place. There were also small fixes to the
  modularization of various system dependencies included within WebKit.
  (Because some of these fixes in WebKit's dependencies have only landed
  in recent SDKs, we do not export the modulemaps when older SDKs
  are in use.)

* Source/WTF/Configurations/WTF.xcconfig:
* Source/WTF/Configurations/modulemap.toml: Added.
* Source/WTF/Scripts/modulemap-from-tapi-filelist.py: Added.
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/Compiler.h:
* Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:
* Source/WebKit/WebKitSwift/IdentityDocumentServices/WKIdentityDocumentPresentmentDelegate.h:
* Source/bmalloc/Configurations/bmalloc.xcconfig:
* Source/bmalloc/Configurations/module.modulemap: Added.
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:

Co-authored-by: Elliott Williams <emw@apple.com>
Canonical link: https://commits.webkit.org/297366@main
